[SIL]/[SILOpt] Thunks and specializations are never public.

huonw opened the pull request:
The current behaviour of (some) open classes means private members are
forced to be public, but this only applies to the real members, not
thunks or specializations associated with them.

Fixes rdar://problem/40738913.

huonw commented:
@swift-ci please test

slavapestov commented on include/swift/SIL/SILFunction.h:
```diff
@@ -430,6 +430,11 @@ class SILFunction
   SILLinkage getEffectiveSymbolLinkage() const {
+    // These are implementations details that don't get special treatment: their
+    // stored linkage is the correct one.
+    if (isThunk() || isSpecialization())
```
I think it would make more sense to turn this into an assertion and fix those places that create thunks no?

huonw replied:
Hm, I'm not sure: this is called in places that have just an arbitrary `SILFunction`:

https://github.com/apple/swift/blob/e8ee93b131d2ad360a8ff02034eb6c29718e411c/lib/IRGen/GenDecl.cpp#L1618-L1619

We could push this condition down to them, but I'm not sure it makes sense. To me, it seems reasonable to say that `getEffectiveSymbolLinkage` is the way to get the true IR/ELF/mach-O linkage, and thus having this split here is the right place.

slavapestov replied:
I'm suggesting that `getClassSubclassScope()` should not be set on thunks. Whoever is creating the thunk's SILFunction is also setting `getClassSubclassScope()`, which seems clearly wrong.

huonw replied:
Ah, I had misunderstood. That sounds good.

huonw commented:
Linux failure (for future reference):

```
20:53:10 FAILED: ../buildbot_linux/foundation-linux-x86_64/Foundation/libFoundation.so 
...
../buildbot_linux/foundation-linux-x86_64/Foundation/Foundation/NSCFDictionary.swift.o:NSCFDictionary.swift.o:$S10Foundation15_NSCFDictionaryCMf: error: undefined reference to '$S10Foundation12NSDictionaryC14getDescription33_8C53A72A73712D943F7602D2573DFCFBLL2ofSSSgyp_tF'
```

jrose-apple commented:
Mm. What about override thunks, for concrete subclasses of generic classes?

```
open class Base<T> {
  open func consume(_: T) {}
}

open class Sub: Base<String> {
  open override func consume(_: String) { print("override") }
}
```

slavapestov commented:
@jrose-apple Override thunks get emitted in each translation unit that has a vtable for a subclass of the class. So you're still relying on the original method having public linkage.

huonw commented:
@swift-ci please test

huonw commented:
@swift-ci please test source compatibility

huonw commented:
@swift-ci please test

@slavapestov I've changed to not setting the subclass scope on thunks, except for signature-optimised thunks, since in that case the original entry point to the code gets marked as a thunk, and thus that needs to have the right subclass scope, so that it remains public if it is `private`.

huonw commented:
@swift-ci please smoke test

slavapestov commented on include/swift/SIL/SILFunction.h:
```diff
@@ -573,13 +573,27 @@ class SILFunction
+  void setThunk(IsThunk_t isThunk, bool forSignatureOptimization = false) {
+    // The _original_ function for a method can turn into a thunk through
+    // signature optimization, meaning it needs to retain its subclassScope
+    assert(getClassSubclassScope() == SubclassScope::NotApplicable ||
```
This seems overkill. If you're intent on keeping this, you should add a new case to IsThunk_t for signature-optimized thunks. A future assertion could also ensure that any thunk other than a signature thunk is shared or private, etc.

huonw committed:
[SIL] Separate thunks from signature optimization from other thunks.

Signature optimization is slightly different to (most) other thunks, in that
it's taking an existing function and turning that into a thunk, rather than
creating a thunk that calls an existing function. These symbols can be public,
etc. and so need to be handled a bit different to other types of thunks.

huonw committed:
[SIL]/[SILOpt] Thunks and specializations shouldn't be connected to their parent class.

The "subclass scope" is meant to represent a connection to a vtable (and how
public something needs to be), for things that end up in class
vtables. Specializations and thunks are mostly internal implementation details
and do not end up there, so subclass scope is not applicable to them. This stops
the thunks and specializations being incorrectly public.

(Note, there are some thunks that _are_ public facing: if a function has its
signature optimized, the original entry point becomes a thunk, and this entry
point is what ends up in vtables etc., so needs to remain around, which means
keeping the same hacks for `private` members of an `open` class.)

Fixes rdar://problem/40738913.

huonw commented:
@swift-ci please smoke test

huonw commented:
@swift-ci please test source compatibility
